File: l00interpreter.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""Tiny bootstrapping interpreter for the first bootstrap stage.

Implements an extremely minimal Forth-like language, used to write
tinyboot1.tbf1.

The theory is that first we ‘compile’ the program by reading through
it to find compile-time definitions and actions, which sets up the
initial state of memory; then we ‘run’ the program by directly
interpreting its text, given that initial state.

"""
import sys, cgitb
cgitb.enable(format='text')

def debug(text):
    sys.stderr.write(text + "\n")

start_address = None
memory = []                  # a list of bytes represented as integers
stack  = []
rstack = []

### Compile-time actions.
# Note that these should leave program_counter pointing after the last
# byte they consume, i.e. they should eat_byte the last byte they
# consume.

program_counter = 0

def current_byte():
    return program[program_counter]

def eat_byte():
    global program_counter
    current_byte = program[program_counter]
    program_counter += 1
    return current_byte

def advance_past_whitespace():
    while program_counter < len(program) and current_byte() in ' \n':
        eat_byte()

def advance_to_whitespace():
    while program_counter < len(program) and current_byte() not in ' \n':
        eat_byte()

def get_token():
    advance_past_whitespace()
    # XXX and on EOF?
    rv = current_byte()
    if rv not in "0123456789'": advance_to_whitespace()
    return rv

def eat_comment():
    comment_start = program_counter
    while eat_byte() != ')': pass
    jump_targets[comment_start] = program_counter

def push_dataspace_label(n):
    return lambda: stack.append(n)

def define(name, action):
    assert name not in run_time_dispatch, name
    run_time_dispatch[name] = action

def dataspace_label():
    "Define a label in data space."
    name = get_token()
    define(name, push_dataspace_label(len(memory)))

def call_function(n):
    def rv():
        global program_counter
        rstack.append(program_counter)
        program_counter = n
    return rv

def define_function():
    name = get_token()
    define(name, call_function(program_counter))

def read_number():
    start = program_counter
    while eat_byte() in '0123456789': pass
    return int(program[start:program_counter])

def literal_byte():
    advance_past_whitespace()
    memory.append(read_number())

def as_bytes(num):
    "Convert a 32-byte number into a little-endian byte sequence."
    return [num & 255, num >> 8 & 255, num >> 16 & 255, num >> 24 & 255]

def literal_word():
    "Compile a little-endian literal 32-byte number into data space."
    advance_past_whitespace()
    memory.extend(as_bytes(read_number()))

def allocate_space():
    advance_past_whitespace()
    memory.extend([0] * read_number())

def set_start_address():
    global start_address
    start_address = program_counter

def nop(): pass

def skip_literal_byte():
    eat_byte()                          # skip the '
    eat_byte()                          # skip the character itself

# We have to find the backwards jump targets for loops while scanning
# forward.  Otherwise we’d have to find them by scanning backwards,
# and you can’t correctly skip comments that way, since comments don’t
# nest.

jump_targets = {}

def start_conditional():
    stack.append(program_counter)

def end_conditional():
    jump_targets[stack.pop()] = program_counter

start_loop = start_conditional

def end_loop():
    jump_targets[program_counter] = stack.pop()

compile_time_dispatch = {
    '(': eat_comment,
    'v': dataspace_label,
    ':': define_function,
    'b': literal_byte,
    '#': literal_word,
    '*': allocate_space,
    '^': set_start_address,
    '[': start_conditional, ']': end_conditional,
    '{': start_loop,        '}': end_loop,
    ' ': eat_byte, '\n': eat_byte,
    "'": skip_literal_byte,
}

for digit in '0123456789': compile_time_dispatch[digit] = read_number

def tbfcompile():
    while program_counter < len(program):
        token = get_token()
        if token in compile_time_dispatch:
            # print "got token at compile time " + str(token)
            compile_time_dispatch[token]()
        elif token in run_time_dispatch:
            pass                 # ignore things from run-time for now
        else:
            excerpt_beginning = max(0, program_counter - 30)
            assert False, '%r not defined at %r' % \
                   (token, program[excerpt_beginning:program_counter])
        # To ensure the loop test condition hits the EOF instead of
        # get_token() choking on it:
        advance_past_whitespace()

### Run-time actions.
# Execution should pretty much stay inside of functions, and we
# shouldn't run into any compile-time actions there, right?

def write_out():
    "Given an address and a count, write out some memory to stdout."
    count = stack.pop()
    address = stack.pop()
    debug('writing address %d, count %d' % (address, count))
    sys.stdout.write(''.join([chr(memory[ii])
                              for ii in range(address, address+count)]))

def quit():
    sys.exit(0)

def subtract():
    x = stack.pop()
    stack.append((stack.pop() - x) & 0xFfffFfff)

def push_literal():
    stack.append(read_number())

def decode(bytes):
    rv = bytes[0] | bytes[1] << 8 | bytes[2] << 16 | bytes[3] << 24
    if rv > 0x7fffFfff:
        rv -= 0x100000000
    return rv

def fetch():
    addr = stack.pop()
    stack.append(decode(memory[addr:addr+4]))

def extend_memory(addr):
    # Addresses > 100k are probably just a bug; it’s not practical to
    # run large programs with this interpreter anyway.
    if len(memory) < addr + 1 and addr < 100000:
        memory.extend([0] * (addr + 1 - len(memory)))

def store():
    addr = stack.pop()
    extend_memory(addr)
    memory[addr:addr+4] = as_bytes(stack.pop())

def store_byte():
    addr = stack.pop()
    extend_memory(addr)
    memory[addr] = stack.pop() & 255

def less_than():
    b = stack.pop()
    a = stack.pop()
    if a < b:
        stack.append(1)
    else:
        stack.append(0)

def return_from_function():
    global program_counter
    program_counter = rstack.pop()

def read_byte():
    byte = sys.stdin.read(1)
    if byte == '':
        stack.append(-1)
    else:
        stack.append(ord(byte))

def jump():
    global program_counter
    program_counter = jump_targets[program_counter]

def conditional():
    if stack.pop(): return
    jump()

def loop():
    if not stack.pop(): return
    jump()

def literal_byte():
    # you put 'A into your program to get 65, or 'B to get 66, etc.
    eat_byte()                          # to skip the '
    stack.append(ord(eat_byte()))

run_time_dispatch = {    
    '(': jump,
    'W': write_out,
    'G': read_byte,
    'Q': quit,
    '-': subtract,
    '<': less_than,
    '@': fetch,
    '!': store,
    # 'f': fetch_byte, not yet needed
    's': store_byte,
    ';': return_from_function,
    '[': conditional, ']': nop,
    '{': nop,         '}': loop,
    ' ': nop, '\n': nop,
    "'": literal_byte,
}

for digit in '0123456789': run_time_dispatch[digit] = push_literal


def tbfrun():
    assert start_address is not None
    global program_counter
    program_counter = start_address
    while True:
        token = get_token()
        # print "got token at run time " + str(token)
        run_time_dispatch[token]()

def main(infile):
    global program
    program = infile.read()
    tbfcompile()
    tbfrun()
    assert False, "tbfrun returned"

if __name__ == '__main__': main(file(sys.argv[1]))

File: l00interpreter.cpp
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <ctype.h>
#include <stdlib.h>
#include <assert.h>

#include <iostream>
#include <unordered_map>
#include <string>
#include <functional>
#include <vector>
#include <stack>

typedef uint8_t word;
typedef std::function<void(void)> vv;

word current_byte();
word eat_byte();
bool is_whitespace(word c);
void advance_past_whitespace();
void advance_to_whitespace();
word get_token();
void eat_comment();
vv push_dataspace_label(uint32_t n);
void define(word name, vv action);
void dataspace_label();
vv call_function(uint32_t n);
void define_function();
uint32_t read_number();
void literal_byte_compile();
std::vector<word> as_bytes(uint32_t n);
void literal_word();
void allocate_space();
void set_start_address();
void nop();
void skip_literal_byte();
void start_conditional();
void end_conditional();
const vv start_loop = start_conditional;
void end_loop();
void tbfcompile();
void write_out();
void quit();
void subtract();
void push_literal();
int32_t decode(std::vector<word> bytes);
void fetch();
void extend_memory(uint32_t addr);
void store();
void store_byte();
void less_than();
void return_from_function();
void read_byte();
void jump();
void conditional();
void loop();
void literal_byte_run();
void tbfrun();

void debug(const char *msg) {
    fprintf(stderr, "%s\n", msg);
}

template<typename T>
T pop(std::stack<T> &v) {
    assert(!v.empty());
    T val = v.top();
    v.pop();
    return val;
}

template<typename K, typename V>
bool intable(K &key, std::unordered_map<K, V> &table) {
    /* These are both correct but I can't decide which is more readable. */
    return table.count(key) > 0;
    return table.find(key) != table.end();
}

std::vector<word> memory;
std::stack<uint32_t> stack;
std::stack<uint32_t> rstack;

std::vector<word> program;
uint32_t pc = 0;
int64_t start_address = -1;

std::unordered_map<uint32_t, uint32_t> jump_targets;

std::unordered_map<word,vv> compile_time_dispatch = {
    {'(', eat_comment},
    {'v', dataspace_label},
    {':', define_function},
    {'b', literal_byte_compile},
    {'#', literal_word},
    {'*', allocate_space},
    {'^', set_start_address},
    {'[', start_conditional}, {']', end_conditional},
    {'{', start_loop}, {'}', end_loop},
    {' ', eat_byte}, {'\n', eat_byte},
    {'\'', skip_literal_byte},
};

std::unordered_map<word, vv> run_time_dispatch = {
    {'(', jump},
    {'W', write_out},
    {'G', read_byte},
    {'Q', quit},
    {'-', subtract},
    {'<', less_than},
    {'@', fetch},
    {'!', store},
    // {'f', fetch_byte}, not yet needed
    {'s', store_byte},
    {';', return_from_function},
    {'[', conditional}, {']', nop},
    {'{', nop}, {'}', loop},
    {' ', nop}, {'\n', nop},
    {'\'', literal_byte_run},
};

word current_byte() {
    return program.at(pc);
}

word eat_byte() {
    return program.at(pc++);
}

bool is_whitespace(word c) {
    return c == ' ' || c == '\n' || c == '\t';
}

void advance_past_whitespace() {
    while (pc < program.size() && is_whitespace(current_byte())) {
        eat_byte();
    }
}

void advance_to_whitespace() {
    while (pc < program.size() && !is_whitespace(current_byte())) {
        eat_byte();
    }
}

word get_token() {
    advance_past_whitespace();
    word rv = current_byte();
    if (!isdigit(rv) && rv != '\'') advance_to_whitespace();
    return rv;
}

void eat_comment() {
    uint32_t comment_start = pc;
    while (eat_byte() != ')') ;
    jump_targets[comment_start] = pc;
}

vv push_dataspace_label(uint32_t n) {
    return [=]() { stack.push(n); };
}

void define(word name, vv action) {
    assert(!intable(name, run_time_dispatch));
    run_time_dispatch[name] = action;
}

void dataspace_label() {
    word name = get_token();
    define(name, push_dataspace_label(memory.size()));
}

vv call_function(uint32_t n) {
    return [=]() {
        rstack.push(pc);
        pc = n;
    };
}

void define_function() {
    word name = get_token();
    define(name, call_function(pc));
}

uint32_t read_number() {
    std::string intbuf;
    char c;
    while (isdigit(c = eat_byte())) {
        intbuf += c;
    }
    return std::stoi(intbuf,nullptr,10);
}

void literal_byte_compile() {
    advance_past_whitespace();
    memory.push_back(read_number());
}

std::vector<word> as_bytes(uint32_t n) {
    std::vector<word> v {(word)n, (word)(n >> 8), (word)(n >> 16), (word)(n >> 24)};
    return v;
}

void literal_word() {
    advance_past_whitespace();
    std::vector<word> bytes = as_bytes(read_number());
    assert(bytes.size() == 4);
    memory.push_back(bytes[0]);
    memory.push_back(bytes[1]);
    memory.push_back(bytes[2]);
    memory.push_back(bytes[3]);
}

void allocate_space() {
    advance_past_whitespace();
    uint32_t n = read_number();
    memory.resize(memory.size() + n, 0);
}

void set_start_address() {
    start_address = pc;
}

void nop() {}

void skip_literal_byte() {
    eat_byte(); // '
    eat_byte(); // char
}

/* */

void start_conditional() {
    stack.push(pc);
}

void end_conditional() {
    jump_targets[pop(stack)] = pc;
}

void end_loop() {
    jump_targets[pc] = pop(stack);
}

void tbfcompile() {
    while (pc < program.size()) {
        word token = get_token();
        if (intable(token, compile_time_dispatch)) {
            compile_time_dispatch[token]();
        }
        else if (intable(token, run_time_dispatch)) {
            ; // ignore things from run-time for now
        }
        else {
            std::string tok = "` '";
            tok[1] = token;
            std::string errmsg = "Illegal instruction encountered at compile time: " + tok;
            debug(errmsg.c_str());
            abort();
        }

        advance_past_whitespace();
    }
}

/* */

void write_out() {
    uint32_t count = pop(stack);
    uint32_t address = pop(stack);
    std::cerr << "writing address " << address << ", count " << count << std::endl;
    std::string buf;
    for (uint32_t i = address; i < address+count; i++) {
        buf += memory[i];
    }
    std::cout << buf;
}

void quit() {
    exit(EXIT_SUCCESS);
}

void subtract() {
    int32_t x = pop(stack);
    stack.push(pop(stack) - x);
}

void push_literal() {
    stack.push(read_number());
}

int32_t decode(std::vector<word> bytes) {
    assert(bytes.size() == 4);
    uint32_t rv = bytes[0] | bytes[1] << 8 | bytes[2] << 16 | bytes[3] << 24;
    return rv;
}

void fetch() {
    uint32_t addr = pop(stack);
    std::vector<word> bytes;
    bytes.push_back(memory[addr+0]);
    bytes.push_back(memory[addr+1]);
    bytes.push_back(memory[addr+2]);
    bytes.push_back(memory[addr+3]);
    stack.push(decode(bytes));
}

void extend_memory(uint32_t addr) {
    /* Address >100k are probably a bug */
    if (memory.size() < addr + 1 && addr < 100000) {
        memory.resize(addr+1, 0);
    }
}

void store() {
    uint32_t addr = pop(stack);
    extend_memory(addr+3); // extend already +1
    std::vector<word> bytes = as_bytes(pop(stack));
    assert(bytes.size() == 4);
    memory[addr+0] = bytes[0];
    memory[addr+1] = bytes[1];
    memory[addr+2] = bytes[2];
    memory[addr+3] = bytes[3];
}

void store_byte() {
    uint32_t addr = pop(stack);
    extend_memory(addr);
    memory[addr] = (word)pop(stack);
}

void less_than() {
    int32_t b = pop(stack);
    int32_t a = pop(stack);
    stack.push((uint32_t) (a<b));
}

void return_from_function() {
    pc = pop(rstack);
}

void read_byte() {
    int byte = getchar();
    if (byte == EOF) {
        debug("EOF");
        stack.push(-1);
    }
    else {
        stack.push(byte);
    }
}

void jump() {
    pc = jump_targets[pc];
}

void conditional() {
    if (pop(stack)) return;
    jump();
}

void loop() {
    if (!pop(stack)) return;
    jump();
}

void literal_byte_run() {
    eat_byte();
    stack.push(eat_byte());
}

void tbfrun() {
    assert(start_address != -1);
    pc = start_address;
    while (true) {
        word token = get_token();
        // std::cout << token << std::endl;
        if (intable(token, run_time_dispatch)) {
            run_time_dispatch[token]();
        }
        else {
            std::string tok = "` '";
            tok[1] = token;
            std::string errmsg = "Illegal instruction encountered at run time: " + tok;
            debug(errmsg.c_str());
            abort();
        }
    }
}

int main(int argc, char **argv) {
    for (char c = '0'; c <= '9'; c++) {
        compile_time_dispatch[c] = read_number;
        run_time_dispatch[c] = push_literal;
    }

    if (argc != 2) {
        debug("Wrong number of arguments");
        return EXIT_FAILURE;
    }

    FILE *fp = fopen(argv[1], "r");
    if (fp == NULL) {
        debug("Could not open file for reading");
        return EXIT_FAILURE;
    }

    int c;
    while ((c = fgetc(fp)) != EOF) {
        program.push_back(c);
    }

    fclose(fp);

    tbfcompile();
    tbfrun();
    abort(); // tbfrun returned -- should exit
}
